### ISSUE TITLE
Support custom config.yaml file for offline environments

### ISSUE THREAD

**Issue by glin** [Bioconductor/BiocManager#75]
Hi, I'm splitting out @slopp's comment about offline use from https://github.com/Bioconductor/BiocManager/issues/71#issuecomment-647802622 into a separate issue.

For offline use, it would be nice if BiocManager supported using a custom `config.yaml` file. Currently, BiocManager always looks up the config file from https://bioconductor.org/config.yaml, which doesn't work in offline environments:
````r
> library(BiocManager)
Warning messages:
1: In file(con, "r") :
  URL 'https://bioconductor.org/config.yaml': status was 'Couldn't resolve host name'
2: In file(con, "r") :
  URL 'http://bioconductor.org/config.yaml': status was 'Couldn't resolve host name'

> BiocManager::repositories()
 Error: Bioconductor version cannot be validated; no internet connection? 
````

To use BiocManager offline, the supported path is to set the `BIOCONDUCTOR_ONLINE_VERSION_DIAGNOSIS` option to `FALSE`, and manually install an appropriate BiocVersion package (from the [vignette](https://cran.r-project.org/web/packages/BiocManager/vignettes/BiocManager.html)):
````r
# Use BiocManager in offline mode
options(BIOCONDUCTOR_ONLINE_VERSION_DIAGNOSIS = FALSE)

# Set custom CRAN mirror and BioC mirror
options(
  repos = "https://my-cran-mirror.org",
  BioC_mirror = "https://my-bioc-mirror.org")
)

# Install BiocManager from CRAN
install.packages("BiocManager")

# Install BiocVersion from manually specified BioC repo with version
install.packages("BiocVersion", repos = "https://my-bioc-mirror.org/packages/3.10/bioc")

BiocManager::version()
# [1] ‘3.10’
````
With `BIOCONDUCTOR_ONLINE_VERSION_DIAGNOSIS=FALSE`, you lose the automatic version detection and possibly other benefits (like CRAN snapshot detection from https://github.com/Bioconductor/BiocManager/issues/71?).

As an alternative to this, could BiocManager support using a custom `config.yaml` file? BioC mirrors could then mirror and host this file for full offline support.

A few ideas on how users could enable this:

1. Configure a custom `BioC_mirror` option, and BiocManager tries to get `config.yaml` from the BioC mirror:
    ````r
    options(BioC_mirror = "https://my-bioc-mirror.org")

    # Look up BioC version from https://my-bioc-mirror.org/config.yaml
    BiocManager::repositories()
    ````
    If the BioC mirror doesn't host the `config.yaml` file, BiocManager falls back to https://bioconductor.org/config.yaml as usual. However, most BioC mirrors don't host `config.yaml` today, so it might be better to opt in to this instead of relying on the fallback.

2. Configure a custom config file URL through an option or environment variable. For example:
    ````r
    # Could be a path to a local file
    options(BIOCONDUCTOR_CONFIG_FILE = "file:///path/to/config.yaml")

    # Or config.yaml from the BioC mirror
    Sys.setenv(BIOCONDUCTOR_CONFIG_FILE = "https://my-bioc-mirror.org/config.yaml")

    # Look up BioC version from BIOCONDUCTOR_CONFIG_FILE
    BiocManager::repositories()
    ````
    This adds another config step, but would still be more convenient than setting the `BIOCONDUCTOR_ONLINE_VERSION_DIAGNOSIS` option.

We'd love to get your thoughts on this - thanks!

**Comment by glin** [comment]
Here's a sample implementation for the first option: https://github.com/Bioconductor/BiocManager/pull/76

**Comment by LiNk-NY** [comment]
This should now be available in `1.30.12`. Thanks!
